Updates for linux install instructions

adegeo opened the pull request:
## Summary

Update all the Linux distro install articles for the projected information. This PR relies on the fact that the feeds are updated. For example, .NET 7 is still not available in the Ubuntu jammy and Kinetic feeds.

I also simplified a lot of the tables and removed outdated information.

Fixes #32521
Fixes #30457
Fixes #32541
Fixes #32663
Fixes #32498

adegeo committed:
Updates for linux install instructions

adegeo committed:
bookmark fix

adegeo committed:
bookmark fix

adegeo commented:
@omajid Can you verify RHEL and the related distros? Thanks!

omajid commented:
Here's the status of .NET packages for Fedora/RHEL-based distros that I know of:

- CentOS
  - Only non-EOL version is CentOS 7, where Microsoft is the only one producing .NET packages for   
- CentOS Stream 8
  - .NET Core 3.1, .NET 6 seem to be available. I expected .NET 7 to be available too, but it's not. I need to look into this. 
- CentOS Stream 9
  - .NET 6 and .NET 7 are available
- Fedora
  - Fedora 35: .NET Core 3.1 and .NET 6 are currently available
  - Fedora 36: .NET Core 3.1 and .NET 6 are currently available
  - Fedora 37: Only.NET 6 is currently available
  - .NET 7 packages are being reviewed for inclusion in Fedora 36 and later. Fedora 35 is being left out because it goes EOL in about a month or so.
- RHEL 7
  - .NET Core 3.1 and .NET 6 are available. .NET 7 is **not** available, and there are no plans to add it.
- RHEL 8
  - .NET Core 3.1, .NET 6 and .NET 7 are all available
- RHEL 9
  - .NET 6 and .NET 7 are available

omajid commented:
I didn't look into the 5.0 packages, let me know if you want me to look into those as well.

omajid commented:
Would it be worth it to include the architectures somewhere? The Microsoft feed only provides .NET packages for `x64`. Some distros provide packages for `arm64` as well. Alpine is the only one that provides packages for `arm` (or is it `arm32`?). Only RHEL 8 and 9 provide `ppc64le` packages.

tdykstra commented on docs/core/install/includes/linux-install-70-dnf.md:
```diff
@@ -0,0 +1,28 @@
+
+### Install the runtime
+
+The ASP.NET Core Runtime allows you to run apps that were made with .NET that didn't provide the runtime. The following command install the ASP.NET Core Runtime, which is the most compatible runtime for .NET. In your terminal, run the following command:
```
```suggestion
The ASP.NET Core Runtime allows you to run apps that were made with .NET that didn't provide the runtime. The following command installs the ASP.NET Core Runtime, which is the most compatible runtime for .NET. In your terminal, run the following command:
```

tdykstra commented on docs/core/install/linux-centos.md:
````diff
@@ -42,24 +38,24 @@ The following table is a list of currently supported .NET releases on both CentO
 ```
 
-[!INCLUDE [linux-yum-install-60](includes/linux-install-60-yum.md)]
+[!INCLUDE [linux-yum-install-60](includes/linux-install-70-yum.md)]
````
```suggestion
[!INCLUDE [linux-yum-install-70](includes/linux-install-70-yum.md)]
```

tdykstra commented on docs/core/install/linux-centos.md:
```diff
@@ -42,24 +38,24 @@ The following table is a list of currently supported .NET releases on both CentO
+.NET Core 3.1 and .NET 5 are available in the default package repositories for CentOS 8.
 
-[!INCLUDE [linux-dnf-install-50](includes/linux-install-50-dnf.md)]
+[!INCLUDE [linux-dnf-install-50](includes/linux-install-31-dnf.md)]
```
```suggestion
[!INCLUDE [linux-dnf-install-31](includes/linux-install-31-dnf.md)]
```

tdykstra commented on docs/core/install/linux-debian.md:
```diff
@@ -18,16 +18,10 @@ This article describes how to install .NET on Debian. When a Debian version fall
-| ❌ [8](#debian-8-)       | ❌ 3.1        | ❌ 6.0   |
+| Debian  | .NET      |
+|---------|-----------|
+| 11      | 7, 6 ,3.1 |
+| 10      | 7, 6 ,3.1 |
```
```suggestion
| 11      | 7, 6, 3.1 |
| 10      | 7, 6, 3.1 |
```

tdykstra commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -20,10 +20,10 @@ The following table is a list of currently supported .NET releases and the versi
-| [20.04 (LTS)](#2004)   | 3.1, 6     |
-| [18.04 (LTS)](#1804)   | 3.1, 6     |
-| [16.04 (LTS)](#1604)   | 3.1, 6     |
+| [22.04 (LTS)](#2210)   | 7, 6       |
```
LInk URL should be #2204 or link text should be 22.10

```suggestion
| [22.10 (LTS)](#2210)   | 7, 6       |
```

tdykstra commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -35,20 +35,69 @@ The following table is a list of currently supported .NET releases and the versi
 
 [!INCLUDE [package-manager uninstall notice](./includes/linux-uninstall-preview-info.md)]
 
+## 22.10
```
The section heading says 22.10, but the "Advisory" link goes to an issue that talks about 22.04

tdykstra commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -135,15 +184,21 @@ sudo apt-get upgrade
+
+Starting with Ubuntu 22.04 you may run into a situation where it seems only a piece of .NET is available. For example, when you've installed the runtime and the SDK, but when running `dotnet --info` it doesn't list the SDK.
+
+This can be related to using two different package sources. The official Ubuntu 22.04 package feeds include .NET and you
```
Incomplete sentence beginning with "The official..."

tdykstra reviewed (approved):
Looks good, I just noted a few typos.

adegeo committed:
Apply suggestions from code review

Co-authored-by: Tom Dykstra <tdykstra@microsoft.com>

adegeo committed:
Feedback from Tom

adegeo commented:
@omajid Thank you for the information. I've updated the articles. For Fedora I removed .NET 7 for the time being but still cleaned things up. Let me know when it's .NET 7 is ready and I'll update the article.

Regarding architecture, it's not something we've really added to any of the install docs yet. We've also not really received much feedback as to confusion about it except when people are trying to install via the MS package manager feed on Ubuntu ARM. Besides that (which the article has a note after that feedback) I've not seen anything else, though I've not been paying much attention to it lately.

adegeo committed:
clarify fedora, cent, rhel

omajid commented:
cc @mirespace to comment on Ubuntu

My understanding is .NET 7 packages in Ubuntu (22.04 and later) aren't quite ready yet.

omajid commented on docs/core/install/linux-alpine.md:
```diff
@@ -23,20 +23,11 @@ Installers aren't available for Alpine Linux. You must install .NET in one of th
 
```
AFAIK, .NET is available in Alpine repositories starting with 3.15?

@ayakael any thoughts on how to update this installation doc?

ayakael replied:
I had actually started working on updating it a few months ago, but lacked the time to complete it. I kind of just copied Fedora's page and updated it for Alpine. .NET 6 was initially included in 3.16, and .NET 7 is slated for 3.17.

adegeo replied:
@omajid @ayakael does someone have instructions for the Alpine package manager? I'm not familiar with that distro in any way. Is it using a manager that we already have instructions for in the include section, such as yum or apt?

omajid replied:
I don't recall the exact install instructions, but to answer the yum/apt part: no, Alpine uses [`apk`](https://wiki.alpinelinux.org/wiki/Alpine_Package_Keeper).

omajid replied:
I tested this using the alpine containers.

```
$ docker pull alpine:3.17
$ docker run -it alpine:3.17
/ # apk add dotnet-sdk-7.0
fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.17/community/x86_64/APKINDEX.tar.gz
ERROR: unable to select packages:
  dotnet-sdk-7.0 (no such package):
    required by: world[dotnet-sdk-7.0]
/ # apk add dotnet-sdk-6.0
(1/17) Installing libgcc (12.2.1_git20220924-r4)
(2/17) Installing libstdc++ (12.2.1_git20220924-r4)
(3/17) Installing dotnet-host (6.0.11-r0)
(4/17) Installing dotnet6-hostfxr (6.0.11-r0)
(5/17) Installing icu-data-full (72.1-r1)
(6/17) Installing icu-libs (72.1-r1)
(7/17) Installing lttng-ust (2.13.3-r0)                                                                                                                       
(8/17) Installing xz-libs (5.2.8-r0)                                           
(9/17) Installing libunwind (1.6.2-r0)
(10/17) Installing dotnet6-runtime (6.0.11-r0)
(11/17) Installing aspnetcore6-runtime (6.0.11-r0)
(12/17) Installing aspnetcore6-targeting-pack (6.0.11-r0)
(13/17) Installing dotnet6-apphost-pack (6.0.11-r0)
(14/17) Installing dotnet6-targeting-pack (6.0.11-r0)
(15/17) Installing dotnet6-templates (6.0.111-r0)
(16/17) Installing netstandard21-targeting-pack (6.0.111-r0)
(17/17) Installing dotnet6-sdk (6.0.111-r0)         
Executing busybox-1.35.0-r27.trigger                                           
OK: 483 MiB in 32 packages
/ # dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.111
 Commit:    b3bb659a9d

Runtime Environment:
 OS Name:     alpine
 OS Version:  3.17
 OS Platform: Linux
 RID:         alpine.3.17-x64
 Base Path:   /usr/lib/dotnet/sdk/6.0.111/

global.json file:
  Not found

Host:
  Version:      6.0.11
  Architecture: x64
  Commit:       943474ca16

.NET SDKs installed:
  6.0.111 [/usr/lib/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.11 [/usr/lib/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.11 [/usr/lib/dotnet/shared/Microsoft.NETCore.App]

Download .NET:
  https://aka.ms/dotnet-download

Learn about .NET Runtimes and SDKs:
  https://aka.ms/dotnet/runtimes-sdk-inf
```

Yeah, just an `apk add <package name>` is sufficient to install available versions of .NET.

ayakael replied:
Indeed, that's the way. dotnet7 is still not merged but is on the way for 3.17 and will be in for main release for, at least, 'armv7`, `aarch64` and `x86_64`. By the way, the package is named `dotnetx-sdk` rather than `dotnet-sdk-x.x`. While we have a virtual package that points to the actual name, we diverged from the guidelines to match our own package naming guidelines.

I've got a working draft of the install instructions for Alpine, and will share when ready.

adegeo replied:
@ayakael what do you mean by 3.16 doesn't support .NET 7 in the table?

omajid commented on docs/core/install/linux-package-mixup.md:
```diff
@@ -1,14 +1,14 @@
 ---
 
-# Troubleshoot _fxr_, _libhostfxr.so_, and _FrameworkList.xml_ errors
+# Troubleshoot .NET errors related to missing files on Linux
```
@mirespace @richlander Do you folks want to add the steps from https://github.com/dotnet/core/issues/7699 to this doc?

omajid replied:
Nevermind, I am an idiot.

omajid commented on docs/core/install/linux-rhel.md:
```diff
@@ -44,13 +41,19 @@ The following table is a list of currently supported .NET releases on both RHEL
+[!INCLUDE [linux-dnf-install-70](includes/linux-install-70-dnf.md)]
 
-## RHEL 7 ✔️ .NET 6.0
+## RHEL 7 ✔️ .NET 6
```
I see the checkmark is removed in other places. Is this just left over?

adegeo replied:
I left here just because it acts as a good visual separator between the acronym and .NET

omajid commented on docs/core/install/includes/linux-install-70-dnf.md:
```diff
@@ -0,0 +1,28 @@
+ms.topic: include
+---
+
+### Install the SDK
```
I might be reading this wrong, but we only refer to this doc (instead of 6.0) when the OS might have multiple .NET versions. Should we tell users how to install .NET 6 here too?

adegeo replied:
This include file has generic steps for installing 7 via dnf. This file is referenced in the linux-rhel article. To keep the articles short and sweet, we only include the latest release's instructions.

adegeo committed:
minor changes; add dep updates

mirespace commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -172,14 +226,20 @@ When you install with a package manager, these libraries are installed for you.
 - libssl1.1 (for 18.x, 20.x)
+- libssl3 (for 22.x)
 - libstdc++6
+- libunwind-13 (for 22.x)
```
We removed the dependency on this (due to the conflict with Discord installation)

mirespace commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -172,14 +226,20 @@ When you install with a package manager, these libraries are installed for you.
 - libicu55 (for 16.x)
 - libicu60 (for 18.x)
 - libicu66 (for 20.x)
+- libicu70 (for 22.04)
```
It would be necessary to add to the list libicu71 (for 22.10)

mirespace commented on docs/core/install/linux-ubuntu.md:
```diff
@@ -35,20 +36,69 @@ The following table is a list of currently supported .NET releases and the versi
+> If you've previously installed .NET from `packages.microsoft.com`, you may run into issues swapping to the built in Ubuntu package manager feeds for .NET. For more information, see the [Advisory on installing .NET on Ubuntu](https://github.com/dotnet/core/issues/7699).
 
-.NET 6 is included in the Ubuntu 22.04 package manager feeds.
+.NET 7 and .NET 6 are included in the Ubuntu 22.04 package manager feeds.
```
Only .NET 6 is included in Ubuntu 22.04 universe repo.

adegeo replied:
.NET 7 isn't going to be published?

lahma replied:
is there an issue or something else to follow progress about when .NET 7 would be available on Ubuntu feeds, if ever?

richlander replied:
We can use this issue to track that: https://github.com/dotnet/core/issues/7939. We can update it as supports gets added. At present, .NET 7 isn't yet in any Ubuntu feed. As @mirespace suggests, .NET 7 is planned for 22.10+.

lahma replied:
So I guess the manual way of adding MS repos and installing from there is the way to go. It's fine, I was just confused whether I should expect the recently announced `apt install` experience to work for Ubuntu LTS.

richlander replied:
Understood. Why I wrote [the post](https://devblogs.microsoft.com/dotnet/dotnet-6-is-now-in-ubuntu-2204/) on that, I didn't know that this would be the plan.

For users that got with LTS only, then this approach will work well. For folks that are on the bleeding edge for everything, it will equally work well. For folks that want LTS Ubuntu + LTS/STS .NET, then it's a pain. In that case, using the MS feeds makes sense.

Separately, .NET 7 isn't yet ready for 22.10. That's just growing pains. The Canonical folks did a great job building and incorporating .NET 6. We (MS) did a poor job of walking Canonical folks through how to manage side-by-side with .NET versions. The delay is due to that. Once we get over the hurdle, we don't have the challenge again.

mirespace commented:
Hi!

.NET 7 is expected to be released next December on 22.10 (*not* on 22.04), for amd64 and arm64.
.NET 6 will be released for arm64, hopefully by the end of this week/beginning of next week in 22.10 and 22.04, along with the amd6 which is already being shipped.

adegeo committed:
more dep changes

adegeo committed:
ubuntu feed doesn't include .net 7

ayakael commented:
> Would it be worth it to include the architectures somewhere? The Microsoft feed only provides .NET packages for `x64`. Some distros provide packages for `arm64` as well. Alpine is the only one that provides packages for `arm` (or is it `arm32`?). Only RHEL 8 and 9 provide `ppc64le` packages.

For now Alpine only provides packages for `aarch64`, `armv7`, `x86_64` and `s390x`. I can send a diff for up-to-date Alpine page, but I'll have to wait a few days as I'm too busy today.

adegeo committed:
Merge branch 'main' of github.adegeo:dotnet/docs into adegeo-linux

adegeo committed:
Merge branch 'main' of github.adegeo:dotnet/docs into adegeo-linux

adegeo committed:
alpine additions

adegeo commented:
@ayakael I ported your diff info, please review and let me know what you think. .NET 6 is still supported on 3.16 correct? Is it just not in the package manager? You marked it as not supportred.

adegeo committed:
add note for ubuntu

ayakael commented:
> @ayakael I ported your diff info, please review and let me know what you think. .NET 6 is still supported on 3.16 correct? Is it just not in the package manager? You marked it as not supportred.

.NET6 is indeed still supported on 3.16. The confusion might stem from the architecture section, where `s390x` support of .NET6 only starts with 3.17. Also, I see that you mark versions on and before 3.15 as supported. This is not the case from Alpine's perspective, as the package was only initially included in 3.16. Should the instructions more explicitely allign with what is supported by Alpine or should they allign more with what is supported by Microsoft?

Also, for what it's worth, I kind of liked the old tables, but I understand the desire to simplify them haha

adegeo commented:
> .NET6 is indeed still supported on 3.16. The confusion might stem from the architecture section, where `s390x` support of .NET6 only starts with 3.17. Also, I see that you mark versions on and before 3.15 as supported. This is not the case from Alpine's perspective, as the package was only initially included in 3.16. Should the instructions more explicitely allign with what is supported by Alpine or should they allign more with what is supported by Microsoft?

Correct, we want to state what is supported from Microsofts standpoint via manual installation, but also point out that there are packages available in the feeds. Does that make sense? Do you think something should be adjusted?

> 
> Also, for what it's worth, I kind of liked the old tables, but I understand the desire to simplify them haha

I also liked the old tables....

ayakael commented:
> @ayakael
> 
> > .NET6 is indeed still supported on 3.16. The confusion might stem from the architecture section, where `s390x` support of .NET6 only starts with 3.17. Also, I see that you mark versions on and before 3.15 as supported. This is not the case from Alpine's perspective, as the package was only initially included in 3.16. Should the instructions more explicitely allign with what is supported by Alpine or should they allign more with what is supported by Microsoft?
> 
> Correct, we want to state what is supported from Microsofts standpoint via manual installation, but also point out that there are packages available in the feeds. Does that make sense? Do you think something should be adjusted?
> 
> > Also, for what it's worth, I kind of liked the old tables, but I understand the desire to simplify them haha
> 
> I also liked the old tables....

Could we bring back the old tables? Alpine would look like this:
```md
- ✔️ indicates that the version of Alpine or .NET is still supported.
- A ❌ indicates that the version of Alpine or .NET isn't supported on that Alpine release.
- When both a version of Alpine and a version of .NET have ✔️, that OS and .NET combination is supported.

| Alpine  | .NET 6 | .NET 7 |
|---------|--------|--------|
| ✔️ 3.17  | ✔️ 6.0  | ✔️ 7.0  |
| ✔️ 3.16  | ✔️ 6.0  | ✔️ 7.0  |
| ✔️ 3.15  | ✔️ 6.0  | ✔️ 7.0  |
| ✔️ 3.14  | ✔️ 6.0  | ❌ 7.0  |

[!INCLUDE [versions-not-supported](includes/versions-not-supported.md)]

## Supported architectures on Alpine package
- ✔️ indicates that the architecture of Alpine or .NET is still supported.
- A ❌ indicates that the architecture of Alpine or .NET isn't supported
- The version number is when the the package was first released for that platform
- Note that only `x86_64`, `armv7`, `aarch64` is officially supported by Microsoft. Other architectures are supported by the distribution maintainers.

| Architecture | .NET 6 | .NET 7 |
|--------------|--------|--------|
| x86_64       | ✔️ 3.16 | ✔️ 3.17 |
| x86          | ❌      | ❌      |
| aarch64      | ✔️ 3.16 | ✔️ 3.17 |
| armv7        | ✔️ 3.16 | ✔️ 3.17 |
| armhf        | ❌      | ❌      |
| s390x        | ✔️ 3.17 | ❌      |
| ppc64le      | ❌      | ❌      |
| riscv64      | ❌      | ❌      |

```

With 3.1 being EOL in a few weeks, I'd propose we remove 3.1 off the tables.

adegeo commented:
Right now the design is to simplify the tables, in line with the rest of the articles. I want to merge this now and then we can follow up with another PR for 3.1 😁 This PR has taken too long due to the lag of everything 😢
